packages/test-utils/src/reporter.ts

```diff
@@ -920,52 +920,106 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
+            failed: false,
+          },
+        };
         this._uploadableResults.set(result.testId, uploadableResults);
       }
-      uploadableResults.push(result);
+      let executions = uploadableResults.executions[result.executionGroupId];
+      if (!executions) {
+        executions = [];
+        uploadableResults.executions[result.executionGroupId] = executions;
+      }
+      executions.push(result);
 
       if (result.result === "passed" || result.attempt >= result.maxAttempts) {
-        this._enqueueUploads(uploadableResults);
+        this._enqueueUploads(uploadableResults, result.executionGroupId);
       }
     }
   }
 
-  private _enqueueUploads(results: UploadableTestRunResult<TRecordingMetadata>[]) {
-    const latestResult = results[results.length - 1];
-    assert(!!latestResult, "Expected at least one result in the list");
+  private _enqueueUploads(
+    result: UploadableTestResult<TRecordingMetadata>,
+    executionGroupId: string
+  ) {
+    const executions = result.executions[executionGroupId];
+    const latestExecution = last(executions);
+    assert(!!latestExecution, "Expected at least one execution in the list");
+
+    let toUpload: typeof executions = [];
 
-    let toUpload: typeof results | undefined;
+    const aggregateStatus = this._assignAggregateStatus(result, executions);
 
-    switch (this._uploadStatusThreshold) {
-      case "all":
-        toUpload = this._minimizeUploads ? [latestResult] : results;
+    switch (aggregateStatus) {
+      case "failed": {
+        if (!this._minimizeUploads) {
+          toUpload.push(...executions);
+          break;
+        }
+        if (result.didUploadStatuses.failed) {
+          break;
+        }
+        result.didUploadStatuses.failed = true;
+        toUpload.push(latestExecution);
         break;
-      case "failed-and-flaky":
-        // retries can be disables so we need to always check if the latest result is not passed
-        // with retries enabled we know that we only have to upload when there was more than one attempt at the test
-        // a single passed attempt can be safely ignored, multiple attempts mean that the test is flaky or failing
-        if (latestResult.result !== "passed" || results.length > 1) {
-          if (!this._minimizeUploads) {
-            toUpload = results;
-          } else if (results.length > 1) {
-            toUpload = [results[0], latestResult];
-          } else {
-            toUpload = [latestResult];
+      }
+      case "flaky": {
+        if (this._uploadStatusThreshold === "failed") {
+          break;
+        }
+        if (!this._minimizeUploads) {
+          result.didUploadStatuses.failed ||= executions.some(r => r.result !== "passed");
+          result.didUploadStatuses.passed ||= executions.some(r => r.result === "passed");
+
+          // currently previously completed execution groups that could be entirely passed or failed are not retroactively uploaded here
```

Comment:
I don't think it's particularly important to "fix" this right now but I could certainly change my mind. 

If we have such execution groups, we detect flake when the third group completes:
```
[passed]
[passed]
[failed, passed]
```

We could look for all completed groups and upload them but it doesn't feel overly important to me.